### ISSUE TITLE
Slow apply_gain kernel in use

### ISSUE THREAD

**Issue by JSKenyon** [ratt-ru/CubiCal#277]
For reasons I do not understand, the performant cyapply_gains function is currently called [cyapply_gains_2](https://github.com/ratt-ru/CubiCal/blob/de4d1162653509cb9e36e2b7451afc10d98b0fca/cubical/kernels/includes/complex_gain_kernel.pxi#L646) and is, in fact, never invoked. The [cyapply_gains code](https://github.com/ratt-ru/CubiCal/blob/de4d1162653509cb9e36e2b7451afc10d98b0fca/cubical/kernels/includes/complex_gain_kernel.pxi#L522) in use is ~7 times slower.

**Comment by o-smirnov** [comment]
Well, on com08 I see this:
```
$ python benchmark/kernel_timings.py cyfull_complex
...
cyfull_complex_reference.apply_gains (DMTFAA order, native)           : 17.11ms (best of 3)
cyfull_complex.cyapply_gains (AAMTFD view)                            : 18.61ms (best of 3)
cyfull_complex.cyapply_gains_1 (AAMTFD view)                          : 44.35ms (best of 3)
cyfull_complex.cyapply_gains_2 (AAMTFD view)                          : 55.56ms (best of 3)
cyfull_complex.cyapply_gains_3 (AAMTFD view)                          : 53.87ms (best of 3)
cyfull_complex.cyapply_gains_slow (AAMTFD view)                       : 55.02ms (best of 3)
cyfull_complex.cyapply_gains_1 (AAMTFD view)                          : 44.00ms (best of 3)
cyfull_complex.cyapply_gains_2 (AAMTFD view)                          : 55.07ms (best of 3)
```
Which machine/scenario are you seeing it as more performat on/in?

**Comment by JSKenyon** [comment]
This was just on my local machine where I was comparing numba and cython. But I have just realized something - this particular kernel somehow has different memory ordering, does it not?

**Comment by JSKenyon** [comment]
It is possible that I have tripped myself up, although it must be said that we really should document input ordering in the docstring, as this is a nasty gotcha.

**Comment by o-smirnov** [comment]
Yeah. The ``kernel_timings.py`` script can be coaxed to benchmark it with different orderings.

Interesting that the results should be so divergent on different hardware. Makes a potential project for auto-tuning all the more interesting. There's a reason why FFTW has a "wisdom" thingy...

When I look at the CPU models on the RATT cluster, there's a large range of L1/L2/L3 cache sizes. So I bet we'll get different results across the cluster as well.

**Comment by o-smirnov** [comment]
> It is possible that I have tripped myself up, although it must be said that we really should document input ordering in the docstring, as this is a nasty gotcha.

But the input ordering is the same always, for all of them. The memory ordering may be different, in which case a view is passed in.

**Comment by JSKenyon** [comment]
Perhaps I can show you when you have a moment - I am still running some tests.

**Comment by sjperkins** [comment]
> When I look at the CPU models on the RATT cluster, there's a large range of L1/L2/L3 cache sizes. So I bet we'll get different results across the cluster as well.

Perhaps this might be a good point to reach back out to the INRIA crowd? They were interested in algorithm auto-tuning. Jon's pytest benchmarks could serve as a valuable starting point. I could put an email together tomorrow...

**Comment by o-smirnov** [comment]
Yeah why not! If we can hand it off to somebody, that'd be great. No
students handy, so let's try to push the collaboration.

Cheers,
Oleg

---
Sent from my phone. Quality of spelling inversely proportional to finger
size.

**Comment by JSKenyon** [comment]
I may have solved this mystery. Copying a view of an array freezes in the view ordering. As a result I do not think that the kernel_timings script is always operating on the orderings one would expect. Checking whether things have gone awry is relatively simple using `view.shape` and `view.base.shape`. After a copy, `copy.base` is `None`. In other words, the underlying memory ordering has been destroyed. @o-smirnov would you mind taking a look to verify if my theory is correct?

**Comment by JSKenyon** [comment]
I have verified that this is definitely what is happening. I replaced the setup function for the apply_gains component of kernel_timings.py so that `inuse_model.base[:] = reference_model.base`. This means that the underlying memory ordering remains unchanged, and the view remains consistent with the expected shape. Results are as follows (and consistent with my other tests) with this change:
```
(numbaenv) jonathan@felix:~/CubiCal/test/benchmark$ python kernel_timings.py 
...
*** Apply 
cyfull_complex_reference.apply_gains (DMTFAA order, native)           : 24.24ms (best of 3)
cyfull_complex.cyapply_gains (AAMTFD view)                            : 68.81ms (best of 3)
cyfull_complex.cyapply_gains_1 (AAMTFD view)                          : 95.33ms (best of 3)
cyfull_complex.cyapply_gains_2 (AAMTFD view)                          : 23.23ms (best of 3)
cyfull_complex.cyapply_gains_3 (AAMTFD view)                          : 22.82ms (best of 3)
cyfull_complex.cyapply_gains_slow (AAMTFD view)                       : 23.27ms (best of 3)
```

**Comment by o-smirnov** [comment]
Aha so I was timing the wrong things then. I see now, apply_gains was special since it operates in-place, so I had to make reference copies of the arrays, and ended up stepping into it...

This is what I get now:
```
cyfull_complex.cyapply_gains (AAMTFD view)                            : 38.21ms (best of 3)
cyfull_complex.cyapply_gains_1 (AAMTFD view)                          : 59.04ms (best of 3)
cyfull_complex.cyapply_gains_2 (AAMTFD view)                          : 13.73ms (best of 3)
cyfull_complex.cyapply_gains_3 (AAMTFD view)                          : 13.68ms (best of 3)
cyfull_complex.cyapply_gains_slow (AAMTFD view)                       : 14.34ms (best of 3)
cyfull_complex.cyapply_gains_1 (AAMTFD view)                          : 58.68ms (best of 3)
cyfull_complex.cyapply_gains_2 (AAMTFD view)                          : 13.79ms (best of 3)
```

**Comment by JSKenyon** [comment]
Fixed by #288.
